### ISSUE TITLE
ymd to pass mismatch for Saral b 

### ISSUE THREAD

**Issue by andreustaylor** [remkos/rads#152]
# identification of cycle/pass from time (ymd)
## apparent erroneous offset for one satellite
- sa only
- sample of recent files shows  
    - lag of approx 4 hours or 5 cycles
 - havent checked anything before 2019

## impact is incorrect time coverage for sa in output from rads2nc
Eg `./rads2nc -S sa --ymd 20190616,20190617`
Processes passes  758 to 787  , but should be 753 to 782.

Spot check by manual dump of last time from files in database:

```
pass 752 20190615_22:47 to 20190615_23:37
pass 753 20190615_23:37 to 20190616_00:28
pass 754 20190616_00:28 to 20190616_01:18  # <<-- should be first pass on 20190616
pass 755 20190616_01:18 to 20190616_02:08
pass 756 20190616_02:08 to 20190616_02:58
pass 757 20190616_02:58 to 20190616_03:49
pass 758 20190616_03:49 to 20190616_04:39  # <<-- actual first pass identified 
pass 759 20190616_04:39 to 20190616_05:29
...
pass 781 20190616_23:06 to 20190616_23:54
pass 782 20190616_23:56 to 20190617_00:46  #<<-- should be last for 20190616
pass 783 20190617_00:46 to 20190617_01:37
pass 784 20190617_01:37 to 20190617_02:27
pass 785 20190617_02:27 to 20190617_03:17
pass 786 20190617_03:17 to 20190617_04:08
pass 787 20190617_04:08 to 20190617_04:58  #<<- actual last pass identified
pass 788 20190617_04:58 to 20190617_05:48
```

## guess:  problem in mission definition config 

https://github.com/remkos/rads/blob/c6371227bcf0e6b80e5de30c07223778c4782701/conf/rads.xml#L4255-L4261

Possible worth a compare to old pass table?
```
../data/tables/sab.pas
b 036 0001  994251177.665462  994254195.528336  994252689.218815 251.074270 2755
```

**Comment by remkos** [comment]
Sorry for the delay in response, and thank you for the feedback.
I can understand where this is coming from, and unfortunately it is because RADS4 is trying to be too smart.

This is what is happening in RADS3:
The data selection runs over the cycle and pass numbers and then is checked against the date range. That is very simple, but if you would not give a cycle and pass range, it will take quite some time for RADS3 to go through all the files and see if there is data in the correct date range. We built findcyclepass.pl in the past to go through an separate ASCII file with pass starts and stops to speed this process. The information you got is correct.

So RADS4 tries to be smarter:
Based on the “orbit scenario” (i.e. basic information like cycle duration, passes per cycle, start of the first cycle) RADS4 estimates the pass numbers. This avoided the need for the separate findcyclepass.pl script and pass files.
Now this is all fine in an exact repeat, but not with a slowly drifting orbit like SARAL. The prediction made in software is not accurate any more.

The interim solution for you:
You can use findcyclepass.pl in conjunction with RADS4. Something like this:<pre>./radsstat4 -S sa -V cycle,pass `findcyclepass.pl sat=sa ymd=20190616,20190617`</pre>

The eventual solution for me to implement:
I will need to add to the “orbit scenario” (which is in the file `rads.xml` as you pointed out above) an extra parameter to indicate how much the orbit is drifting. There is already such a parameter to indicate the drift in longitude (-2.30), but not for the change in the length of the passes (as the orbit keeps on dropping). But a regular update of this may be sufficient. Indeed I just looked into the `sab.pas` file (used by `findcyclepass.pl`) and one can see that the cycle during Phase B is significantly longer than 35 days: about 9m30s longer at the start, and now still 8m10s longer (as the orbit lowered). Some average length may be sufficient.
I will surely work on fixing this in due course.

**Comment by andreustaylor** [comment]
Good to get this as a known bug.

I suppose the solution would want to deal with any platform in a drifting orbit.
Looks like 7 such phases listed in rads.xml  `grep -i "drift phase" ../data/conf/rads.xml ` 

In the meantime it may be helpful to:
  1. confirm that the `../data/tables/` files are maintained  
  2. add `findcyclepass.pl` (or equivalent) to this repo somewhere

**Comment by remkos** [comment]
This is now solved since [version (v4.3.6)](https://github.com/remkos/rads/releases/tag/v4.3.6)
